[HACK] Migrate WooLogViewerScreen to Material 3

kidinov opened the pull request:
### Description
Migrated the WooLogViewerScreen from Material 2 to Material 3 components to align with the app's design system modernization. This change updates the log viewer screen to use Material 3 Scaffold, TopAppBar, and other components while maintaining the same functionality and visual appearance.

Key changes:
- Updated WooLogViewerActivity to use ComponentActivity with edge-to-edge support
- Converted all Material 2 components to Material 3 equivalents
- Toolbar uses material 3 now

### Steps to reproduce
1. Open the WooCommerce app
2. Navigate to Settings > Help & Support > Application logs
3. Verify the screen displays correctly in both light and dark modes
4. Test toolbar interactions (back button, copy, share)
5. Test search functionality with navigation arrows
6. Verify edge-to-edge display behavior

### Testing information
- Test on devices with different screen sizes and densities
- Verify proper window insets handling with system bars
- Test orientation changes
- Verify accessibility features still work correctly
- Test with different system themes (light/dark)

### The tests that have been performed
- Visual verification of Material 3 styling
- Functional testing of all toolbar actions
- Search functionality testing
- Dark mode compatibility testing
- Edge-to-edge display verification

### Images/gif

https://github.com/user-attachments/assets/437e14c1-e186-4bf7-af43-938a51308b53



- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

kidinov committed:
Implement Material3 theme support in log viewer components

kidinov committed:
Add Material3 Toolbar component and update log viewer to use it

kidinov committed:
Refactor log viewer to use ComponentActivity and update toolbar colors

kidinov committed:
Update ToolbarMaterial3 to use MaterialTheme typography for title

kidinov committed:
Remove unused WindowInsets imports and simplify padding in WooLogViewerScreen

kidinov committed:
Enable edge-to-edge display for Material 3 log viewer screen

Added enableEdgeToEdge() call to WooLogViewerActivity to ensure proper
edge-to-edge behavior with Material 3 components. This allows the
Material 3 Scaffold and TopAppBar to handle window insets automatically
while providing seamless display behind system bars.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

kidinov committed:
Remove accidentally committed files

Removed .claude/commands/create-pr.md and 07-30--09-56.mp4 from version control.
These files were not intended to be part of the Material 3 migration.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/theme/ThemeMaterial3.kt:
```diff
@@ -0,0 +1,120 @@
+ * @see <a href="https://developer.android.com/jetpack/compose/themes/material#content-color</a> for more details
+ */
+@Composable
+fun WooThemeM3WithBackground(
```
@kidinov there is no need for defining a new theme, we just added a Material 3 theme definition to our `WooTheme` declaration, it was just merged today (https://github.com/woocommerce/woocommerce-android/pull/14394), so you can just rebase your branch onto trunk, and remove these new declarations, and then continue to use `WooThemeWithBackground` as before, it will supply the correct colors.

kidinov replied:
Didn't know that. Using that theme now.

Btw, are you sure that it's a good idea to wrap one theme into another?

hichamboushaba replied:
> Btw, are you sure that it's a good idea to wrap one theme into another?

Yes, it should be fine, the themes provide composition locals for different classes (for example: `androidx.compose.material.LocalShapes` vs `androidx.compose.material3.LocalShapes`, and `androidx.compose.material.LocalColors` vs `androidx.compose.material3.LocalColorScheme`), so they don't conflict with each other. The themes will override two common composition locals: `LocalIndication` and `LocalTextSelectionColors`, for those, the last one will override the other one, but from my testing, both themes define almost the same colors for those.

Also, from my testing and Irfan's testing when reviewing the PR, this approach didn't cause any issues. But if you have any concerns, please share them 🙏.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/ToolbarMaterial3.kt:
```diff
@@ -0,0 +1,84 @@
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun ToolbarM3(
```
Just asking a question here, WDYT about updating the current `Toolbar` to use the Material 3 components under the hood instead of adding a new component?

This is the approach I followed in this PR https://github.com/woocommerce/woocommerce-android/pull/14401, and to me, this will make the migration easier, as the screens that use the components will be migrated correctly, and with the correct theming, the components will look the same, so no UI impact.

kidinov replied:
I tried that and seems like it works fine

0c8e97f2f73ee1c8626a3bb2fee85a4cdf2a3579

kidinov committed:
Refactor to use WooThemeM3 instead of WooThemeM3WithBackground and remove unused WindowInsets imports

kidinov committed:
Update background color in SearchLayoutWithParams to use toolbar color

kidinov committed:
Show copy and share buttons only when search query is empty in WooLogViewerScreen

kidinov committed:
Add contentWindowInsets to fillMaxSize modifier in WooLogViewerScreen

kidinov committed:
Add horizontal window insets padding to fillMaxSize modifier in WooLogViewerScreen

kidinov committed:
Merge branch 'trunk' into hack-material3-app-logs

kidinov committed:
Replace WooThemeM3 with WooTheme in log viewer components

kidinov committed:
Replace painterResource with imageVector for copy and share icons in WooLogViewerScreen

kidinov committed:
Refactor Toolbar components to use Material3 styles and update window insets

kidinov committed:
Update icon tint colors in WooLogViewerScreen to use primary color from MaterialTheme

kidinov committed:
Imports ordering

kidinov committed:
Release notes update

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/Toolbar.kt:
```diff
@@ -44,40 +47,42 @@ fun ToolbarWithHelpButton(
     Toolbar(
         modifier = modifier,
-        title = { Text(title) },
+        title = { Text(title, style = MaterialTheme.typography.headlineSmall) },
```
```suggestion
        title = { Text(title) },
```

The `headlineSmall` style is causing a significant change from the old implementation, and thus causing inconsistencies with other screens:

| Before | After |
| --- | --- |
| <img width="1080" height="2400" alt="Screenshot_20250731_145419" src="https://github.com/user-attachments/assets/613a484a-8da9-4da1-920d-d51879313d37" /> | <img width="1080" height="2400" alt="Screenshot_20250731_145317" src="https://github.com/user-attachments/assets/383a9df7-6fe1-423a-853f-9374e20e4d2c" /> |

I tested removing the explicit style to use the default style provided by the `TopAppBar`, and it seems to be closer to what had before:

<img width="360" alt="Screenshot_20250731_145728" src="https://github.com/user-attachments/assets/ac986060-fa80-4e6d-9681-d95bd276ea05" />

kidinov replied:
Good catch!

51a9bb178df2bf9870f7273d7b1e2e25f46744cb

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/Toolbar.kt:
```diff
@@ -44,40 +47,42 @@ fun ToolbarWithHelpButton(
     Toolbar(
         modifier = modifier,
-        title = { Text(title) },
+        title = { Text(title, style = MaterialTheme.typography.headlineSmall) },
```
Same as above

```suggestion
        title = { Text(title) },
```

kidinov replied:
51a9bb178df2bf9870f7273d7b1e2e25f46744cb

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/Toolbar.kt:
```diff
@@ -86,26 +91,28 @@ fun Toolbar(
                     imageVector = actionButtonIcon,
-                    contentDescription = actionIconContentDescription
+                    contentDescription = actionIconContentDescription,
+                    tint = MaterialTheme.colorScheme.onSurface
```
I don't think we need this `tint` override here, `TopAppBar` will tint the icons using the `actionIconContentColor` we are providing below.

kidinov replied:
Removed redundant redeclaration

fdbeaca6010ec99fdb0baf33c359053b9320e9fd

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/Toolbar.kt:
```diff
@@ -140,19 +148,25 @@ fun Toolbar(
+            containerColor = colorResource(id = R.color.color_toolbar),
+            titleContentColor = MaterialTheme.colorScheme.onSurface,
+            navigationIconContentColor = MaterialTheme.colorScheme.onSurface,
+            actionIconContentColor = MaterialTheme.colorScheme.onSurface
```
We use primary color for Toolbar actions (you can confirm it in other app screens)
```suggestion
            actionIconContentColor = MaterialTheme.colorScheme.primary
```

kidinov replied:
Good catch!

1f1aaa753919f0363e6bce084ff977e79db26fbc

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/Toolbar.kt:
```diff
@@ -2,33 +2,36 @@ package com.woocommerce.android.ui.compose.component
     modifier: Modifier = Modifier,
     title: String = "",
     onNavigationButtonClick: (() -> Unit)? = null,
     navigationIcon: ImageVector? = Icons.AutoMirrored.Filled.ArrowBack,
     navigationIconContentDescription: String = stringResource(id = string.back),
-    windowInsets: WindowInsets = WindowInsets(0),
+    windowInsets: WindowInsets = TopAppBarDefaults.windowInsets,
```
The default insets we get from this change is `WindowInsets.systemBarsForVisualComponents.only(WindowInsetsSides.Horizontal + WindowInsetsSides.Top)`,This change have some side effects depending on where the Toolbar is used and whether the insets were consumed or not, an example where this breaks the app is screens that use the Toolbar and the keyboard is shown:

<img width="360" alt="Screenshot_20250731_151923" src="https://github.com/user-attachments/assets/5b43b50a-680f-4270-945e-a41cdcd7c609" />
<img width="360" alt="Screenshot_20250731_152109" src="https://github.com/user-attachments/assets/47cf4170-bf4a-4a92-9d59-d618b2ee62db" />

I think we should revert to `WindowInsets(0)` for now, and then update the `WooLogViewerScreen` to either revert to applying the `WindowInsetsSides.Top` direction of the padding, or pass a specific inset to the Toolbar.

kidinov replied:
Interestingly, I cannot reproduce this on my emulator. I tried both SDK 35 and 33. What device/emulator you use? 

<img width="327" height="730" alt="image" src="https://github.com/user-attachments/assets/c8f1a730-a20a-4b18-955b-fa354f1ee02a" />

<img width="331" height="730" alt="image" src="https://github.com/user-attachments/assets/212e369b-14bf-40f0-92c6-a2902f209a2a" />

<img width="325" height="708" alt="image" src="https://github.com/user-attachments/assets/e00338a9-8f4a-4702-b0f0-45f9f74894eb" />


What I really don't like about `windowInsets: WindowInsets = WindowInsets(0),` is that we override the default behavior of the Material 3 TopAppBar, which is supposed to handle edge-to edge out of the box

hichamboushaba replied:
Thanks @kidinov for testing, and sorry for not making it clearer (I mentioned it above, but it was easy to miss). To reproduce the issue, you need to trigger displaying the keyboard, then hide it in one of the mentioned screens. Did you do this?

After looking more into this, I see it is related to this [issue](https://issuetracker.google.com/issues/411868840) that I reported to Google, which seems to have been fixed recently, and will be [released](https://developer.android.com/jetpack/androidx/releases/compose-foundation#1.9.0-alpha02) with the upcoming Compose 1.9, but for now, we'll have to avoid using consumed insets in Compose views.

> What I really don't like about windowInsets: WindowInsets = WindowInsets(0), is that we override the default behavior of the Material 3 TopAppBar, which is supposed to handle edge-to edge out of the box

I understand this, but I think the default that the team uses is more geared towards full Compose apps, for apps that still use the View system as the root layout (like what we do), we have some insets handling in our layouts which collides with the Compose system, and this is what happens here, that's why I think using `WindowInsets(0)` like the default that was in Material 2 still works best for our app, and given it's applied on a Component, we can easily update it in the future when the bug is fixed.

kidinov replied:
@hichamboushaba 

Make sense!

Implemented here f9dda37b00926578d7ba497054da538d31a964f4

hichamboushaba commented on RELEASE-NOTES.txt:
```diff
@@ -4,13 +4,14 @@
 23.0
 -----
+- [*] Add search functionality to application log viewer [https://github.com/woocommerce/woocommerce-android/pull/14386]
+- [*] Edge to edge support to application log viewer [https://github.com/woocommerce/woocommerce-android/pull/14406]
```
Just wondering, what part was broken here? 
From comparing the before and after, all I can notice is that the background doesn't expand to edge-to-edge anymore, IMO both approaches are valid.

| Before | After |
| --- | --- |
| <img width="2400" height="1080" alt="Screenshot_20250731_151252" src="https://github.com/user-attachments/assets/ac6d0838-7a3e-4f1f-99b2-86064ffd9d43" /> | <img width="2400" height="1080" alt="Screenshot_20250731_151139" src="https://github.com/user-attachments/assets/a76247f5-bbc0-485f-b312-266265ea9d73" /> |

kidinov replied:
I am sure that it had bottom padding taht was covering the navigation bar, but now I tested on trunk and even on 22.8 and it's not there. So weird =). I removed that from the release notes then

0c9edf7ddb9300a49f2c64ce8537e002c3aec78c

hichamboushaba replied:
Thank you 🙏

hichamboushaba reviewed (changes requested):
Thanks @kidinov, great work here, but there are some issues that need fixing before merging it.

kidinov committed:
Remove custom typography style from Toolbar title

kidinov committed:
Remove tint color from action button icon in Toolbar

kidinov committed:
Merge branch 'trunk' into hack-material3-app-logs

kidinov committed:
Updated release notes

kidinov committed:
Update icon tint colors in log viewer and toolbar

kidinov committed:
Refactor windowInsets handling in Toolbar and update WooLogViewerScreen for Material3 compatibility

hichamboushaba committed:
Opt-in to the experimental API

hichamboushaba reviewed (approved):
Nice work @kidinov, thanks for taking the time to fix the issues.

I took the liberty to push one last commit to fix the compilation issue, I hope that's fine.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/264">23.0</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
